pkg/gitprovider/gogs.go

```diff
@@ -0,0 +1,282 @@
+	// Gogs does not have any API endpoint to fetch PRs
+	return []*GitPullRequest{}, nil
+}
+
+func (g *GogsGitProvider) GetBranchByCommit(staticContext *StaticGitContext) (string, error) {
+	client := g.getApiClient()
+	repoBranches, err := client.ListRepoBranches(staticContext.Owner, staticContext.Name)
+	if err != nil {
+		return "", fmt.Errorf("failed to get branch by commit: %w", err)
+	}
+
+	var branchName string
+	for _, branch := range repoBranches {
+		commitId := branch.Commit.ID
+		if *staticContext.Sha == commitId {
+			branchName = branch.Name
+			break
+		}
+
+		for commitId != "" {
+			commit, err := client.GetSingleCommit(staticContext.Owner, staticContext.Name, commitId)
+			if err != nil {
+				continue
+			}
+
+			if *staticContext.Sha == commit.SHA {
+				branchName = branch.Name
+				break
+			}
+			if len(commit.Parents) > 0 {
+				commitId = commit.Parents[0].SHA
+				if *staticContext.Sha == commitId {
+					branchName = branch.Name
+					break
+				}
+			} else {
+				commitId = ""
+			}
+		}
+
+		if branchName != "" {
+			break
+		}
+	}
+
+	if branchName == "" {
+		return "", fmt.Errorf("status code: %d branch not found for SHA: %s", http.StatusNotFound, *staticContext.Sha)
+	}
+	return branchName, nil
+}
+
+func (g *GogsGitProvider) GetLastCommitSha(staticContext *StaticGitContext) (string, error) {
+	client := g.getApiClient()
+	branchName := ""
+	if staticContext.Branch != nil {
+		branchName = *staticContext.Branch
+	}
+	branch, err := client.GetRepoBranch(staticContext.Owner, staticContext.Name, branchName)
+	if err != nil {
+		return "", fmt.Errorf("failed to get last commit SHA: %w", err)
+	}
+	return branch.Commit.ID, nil
+}
+
+func (g *GogsGitProvider) ParseStaticGitContext(repoUrl string) (*StaticGitContext, error) {
+	staticContext, err := g.AbstractGitProvider.ParseStaticGitContext(repoUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	if staticContext.Path == nil {
+		return staticContext, nil
+	}
+
+	parts := strings.Split(*staticContext.Path, "/")
+	switch {
+	case len(parts) >= 2 && parts[0] == "pulls":
+		prNumber, err := strconv.Atoi(parts[1])
+		if err != nil {
+			return nil, err
+		}
+		prUint := uint32(prNumber)
+		staticContext.PrNumber = &prUint
+		staticContext.Path = nil
+	case len(parts) >= 2 && parts[0] == "src":
+		staticContext.Branch = &parts[1]
+		if len(parts) > 2 {
+			branchPath := strings.Join(parts[2:], "/")
+			staticContext.Path = &branchPath
+		} else {
+			staticContext.Path = nil
+		}
+	case len(parts) >= 2 && parts[0] == "commits":
+		staticContext.Branch = &parts[1]
+		staticContext.Path = nil
+	case len(parts) >= 2 && parts[0] == "commit":
+		staticContext.Sha = &parts[1]
+		staticContext.Branch = staticContext.Sha
+		staticContext.Path = nil
+	}
+
+	return staticContext, nil
+}
+
+func (g *GogsGitProvider) GetPrContext(staticContext *StaticGitContext) (*StaticGitContext, error) {
```

Comment:
We should throw an error if the user attempts to create a workspace from a PR.
![Screenshot 2024-10-16 at 13 40 01](https://github.com/user-attachments/assets/9ddfad24-2e82-40f1-abfb-bf3884ae420d)
![Screenshot 2024-10-16 at 13 39 58](https://github.com/user-attachments/assets/9b5dcb89-5b0d-4152-8642-da2c2ff08acd)

I created a fork of a repo and submitted a PR. When I used the PR to create a workspace, it attached the wrong context (of the target repo instead of the source) so the cloned repo inside is wrong.

This is not very useful so I suggest that we throw an appropriate error instead.

Reply:
done